### PR TITLE
Fix system module with both filesets enabled

### PR DESCRIPTION
**Pull request by belimawr**
## Proposed commit message

The system module did not define an ID at the root of the config, that made the V2 input loader only start the first journald input it saw because they both ended up with the same identifier (type, ID and path). This is fixed by defining an ID at the root of the configuration templates.

The journald input now also adds the `input_id` key to its loggers and a non-fatal error is now logged at debug level.

The system-logs input is now marked as experimental instead of stable.

Fix lint warnings by moving `toJournalConfig` to `input_linux.go`

Move TestSystemLogsCanUseLogInput to a file without the
linux build constraint so it can run on all OSes supported by the
system integration.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

~~## Disruptive User Impact~~
~~## Author's Checklist~~

## How to test this PR locally

1. Run Filebeat with the following `filebeat.yml` and `modules.d/system.yml` (adjust credentials/addresses as necessary)
    <details><summary>filebeat.yml</summary>
    <p>
    
    ```yaml
    filebeat.inputs:
      - type: journald
        id: my-journald-normal-input
        tags:
           - journald-input
      - type: filestream
        id: my-filestream-id
        paths:
          - /tmp/flog.log
    
    filebeat.config.modules:
      path: ${path.config}/modules.d/*.yml
      reload.enabled: false
      reload.period: 1s
    
    setup.template:
      settings:
        index.number_of_shards: 1
    
    setup.kibana:
      host: "http://kibana:5601"
      username: admin
      password: testing
      ssl.verification_mode: none
    
    output.elasticsearch:
      hosts: ["http://elasticsearch:9200"]
      preset: latency
      protocol: "http"
    
      username: admin
      password: testing
      ssl.verification_mode: none
    ```
    
    </p>
    </details> 
    
    <details><summary>modules.d/system.yml</summary>
    <p>
    
    ```yaml
    - module: system
      syslog:
        enabled: true
        var.use_journald: true
        input:
          tags:
            - from-journald
    
      auth:
        enabled: true
        var.use_journald: true
        var.tags:
          - from-journald
    ```
    
    </p>
    </details> 

2. Go to Discover in Kibana, filter by `tags: from-journald`
3. Look at `fileset.name` from the events, make sure `auth` and `syslog` are there

## Related issues
- Closes https://github.com/elastic/beats/issues/41378

~~## Use cases~~
~~## Screenshots~~
~~## Logs~~

### CODE REVIEW

File: filebeat/input/journald/pkg/journalctl/journalctl.go [hunk 1]
```diff
@@ -95,7 +96,20 @@ func Factory(canceller input.Canceler, logger *logp.Logger, binary string, args
 			data, err := reader.ReadBytes('\n')
 			if err != nil {
 				if !errors.Is(err, io.EOF) {
-					logger.Errorf("cannot read from journalctl stdout: %s", err)
+					var logError = false
+					var pathError *fs.PathError
+					if errors.As(err, &pathError) {
+						if pathError.Op == "read" && pathError.Path == "|0" && pathError.Err.Error() == "file already closed" {
```
Comment (mauri870) [inline filebeat/input/journald/pkg/journalctl/journalctl.go:102]: Out of curiosity, what path `"|0"` means?

File: filebeat/tests/integration/systemlogs_test.go [hunk 2]
```diff
@@ -63,6 +68,83 @@ func TestSystemLogsCanUseJournaldInput(t *testing.T) {
...
+				t.Fatalf("cannot parse output line as JSON: %s", err)
+			}
+
+			if data.Fileset.Name == "syslog" {
+				foundSyslog = true
+			}
+
+			if data.Fileset.Name == "auth" {
+				foundAuth = true
+			}
```
Comment (mauri870) [inline filebeat/tests/integration/systemlogs_test.go:137]: I think a switch might be more readable here and easier to extend (if that is relevant):

```go
switch data.Fileset.Name {
case "syslog":
    foundSyslog = true
case "auth":
    foundAuth = true
}
```

File: filebeat/tests/integration/testdata/filebeat_system_module.yml [hunk 3]
```diff
@@ -14,3 +18,4 @@ output:
   file:
     path: ${path.home}
     filename: "output"
+    rotate_every_kb: 62500 # 500Mb
```
Comment (mauri870) [inline filebeat/tests/integration/testdata/filebeat_system_module.yml:21]: Is this every 62500kb? If so that is 62,5MB and not 500. Did I miss something?

File: filebeat/tests/integration/systemlogs_test.go [hunk 4]
```diff
@@ -63,6 +68,83 @@ func TestSystemLogsCanUseJournaldInput(t *testing.T) {
...
+				"only a single output file is supported, found: %d. Files: %s",
+				len(files),
+				files,
+			)
+		}
+
+		foundSyslog := false
+		foundAuth := false
+
+		file, err := os.Open(files[0])
```
Comment (rdner) [inline filebeat/tests/integration/systemlogs_test.go:108]: The linter says the `err` value is not used below.

File: filebeat/input/journald/pkg/journalctl/journalctl.go (hunk 1 above)
Comment (belimawr) [reply filebeat/input/journald/pkg/journalctl/journalctl.go:102]: I believe, based on observations, it means `stdout`. I searched a bit for it in the Go source code but did not find many references. It appears a couple of times in the `os` package:
- https://github.com/golang/go/blob/d0631b90a3b0934d0fe223e2d889d785d297f083/src/os/pipe_unix.go#L27
- https://github.com/golang/go/blob/d0631b90a3b0934d0fe223e2d889d785d297f083/src/os/pipe2_unix.go#L21

I did not try reproducing this situation and tracing back how the error gets constructed. Those values for `fs.PathError` are based on observations of what happens. I believe I can add some comments there better explaining the situation.

In short, when reading from journalctl stdout, sometimes journalctl will successfully exit and we're still trying to read from it, sometimes we get a EOF, other times we get this error. I believe it is caused by a timing 'issue' while the process is being shutdown and we're still trying to read. It is harmless and no data is lost, but we have to handle the error.

I ended up adding this condition there to avoid having a log level error on user's deployments that are an expected and well handled situation.

File: filebeat/tests/integration/systemlogs_test.go (hunk 2 above)
Comment (belimawr) [reply filebeat/tests/integration/systemlogs_test.go:137]: Thanks! Indeed it's much better :D

File: filebeat/tests/integration/systemlogs_test.go (hunk 4 above)
Comment (belimawr) [reply filebeat/tests/integration/systemlogs_test.go:108]: Thanks!

File: filebeat/tests/integration/testdata/filebeat_system_module.yml (hunk 3 above)
Comment (belimawr) [reply filebeat/tests/integration/testdata/filebeat_system_module.yml:21]: I didn't pay attention to the bit/byte difference there :man_facepalming:

File: filebeat/input/journald/pkg/journalctl/journalctl.go (hunk 1 above)
Comment (mauri870) [reply filebeat/input/journald/pkg/journalctl/journalctl.go:102]: Oh, yeah it makes sense now. It is a pipe to stdout, kinda obvious when you reason with each symbol individually, but for some reason it did not click for me at first. TIL. 

A short comment explaining that this is checking specifically for read errors from the journald stdout would be great.

File: filebeat/tests/integration/testdata/filebeat_system_module.yml (hunk 3 above)
Comment (mauri870) [reply filebeat/tests/integration/testdata/filebeat_system_module.yml:21]: It happens to the best of us! Thanks for addressing.

**Review by mauri870 — COMMENTED** [review]
I don't see the linter change nor the comment for the journald stdout, did you forgot to push?

**Review by mauri870 — APPROVED** [review]

**Comment by belimawr** [comment]
Some CI failures seem to be unrelated:
```
<html>
<body>
<!--StartFragment-->
=== Failed
--
  | === FAIL: filebeat/input/filestream/internal/task TestGroup_Go/workloads_wait_for_available_worker (0.20s)
  | group_test.go:129: f2 started
  | group_test.go:116: f1 started
  | group_test.go:135: f2 done
  | group_test.go:148: f3 started
  | group_test.go:182: waiting the worker pool to finish all workloads
  | group_test.go:118: f1 done
  | group_test.go:150: f3 done
  | group_test.go:185: worker pool to finished all workloads
  | group_test.go:187:
  | Error Trace:	C:/buildkite-agent/builds/bk-agent-prod-gcp-1729720802417888488/elastic/filebeat/filebeat/input/filestream/internal/task/group_test.go:187
  | Error:      	Condition never satisfied
  | Test:       	TestGroup_Go/workloads_wait_for_available_worker
  | Messages:   	not all goroutines finished
  |  
  | === FAIL: filebeat/input/filestream/internal/task TestGroup_Go (0.70s)

<!--EndFragment-->
</body>
</html>
```
BK link: https://buildkite.com/elastic/filebeat/builds/10493#0192bb64-7d3a-4e8a-9879-e4528b7076ab/154-388

I tried to reproduce it on Win and did not succeed. I'll just re-run, if it fails again I'll open a FlakyTest issue.

**Comment by mauri870** [comment]
@belimawr This test looks suspiciously like the other ones I fixed recently https://github.com/elastic/beats/pull/41230. I wouldn't be surprised if it is flaky.

**Comment by belimawr** [comment]
> @belimawr This test looks suspiciously like the other ones I fixed recently #41230. I wouldn't be surprised if it is flaky.

I agree! However I didn't manage to reproduce it :/

That's way I posted the message here, it's a reminder of the error and links in case it happens again. ;)

**Comment by mauri870** [comment]
> I agree! However I didn't manage to reproduce it :/

I forgot to reply, but I also tried to reproduce with `x/tools/cmd/stress` but had no luck.

File: .buildkite/filebeat/filebeat-pipeline.yml [hunk 5]
```diff
@@ -111,6 +111,38 @@ steps:
...
+              format: "junit"
+              branches: "main"
+              debug: true
+        notify:
+          - github_commit_status:
+              context: "filebeat: Go Integration Tests"
+
+      - label: ":ubuntu: Filebeat: Integration tests as Root"
+        command: |
+          sudo usermod -a -G systemd-journal $$USER
```
Comment (alexsapran) [inline .buildkite/filebeat/filebeat-pipeline.yml:129]: if i am not wrong, this is how bk escapes the env vars.

```suggestion
          sudo usermod -a -G systemd-journal \$USER
```

Comment (belimawr) [reply .buildkite/filebeat/filebeat-pipeline.yml:129]: Both are supported according to the [docs](https://buildkite.com/docs/pipelines/environment-variables#runtime-variable-interpolation).

`$$USER` worked in the last run, then it failed because I didn't compile the test binary :man_facepalming: .

File: filebeat/integ-test-buildkite.sh
Comment (alexsapran) [inline filebeat/integ-test-buildkite.sh:1]: This file is no longer needed since I adjusted our VM images.

File: filebeat/tests/integration/systemlogs_linux_test.go [hunk 6]
```diff
@@ -50,10 +43,13 @@ func TestSystemLogsCanUseJournaldInput(t *testing.T) {
 
 	// As the name says, we want this folder to exist bu t be empty
 	globWithoutFiles := filepath.Join(filebeat.TempDir(), "this-folder-does-not-exist")
-	yamlCfg := fmt.Sprintf(systemModuleCfg, globWithoutFiles, workDir)
```
Comment (alexsapran) [inline filebeat/tests/integration/systemlogs_linux_test.go:53]: I noticed that on my Draft PR, when I was testing the VM image, this function `TestSystemLogsCanUseJournaldInput` failed and, with a retry, worked. 

So, this is going to be marked as flaky.

Comment (belimawr) [reply filebeat/tests/integration/systemlogs_linux_test.go:53]: That's the big puzzle I'm trying to solve, why it only fails on CI.

**Review by rowlandgeoff — APPROVED** [review]
BK pipeline change looks good
